### ISSUE TITLE
perplexity-val values are incorrect

### ISSUE THREAD

**Issue by xingniu** [awslabs/sockeye#185]
Version: 1.10.1
"perplexity-val" values in model/metrics are incorrect, e.g.
```
1	avg-sec-per-sent-val=0.406752	bleu-val=0.069343	perplexity-train=162.859366	perplexity-val=121212.751628	time-elapsed=356.103674	used-gpu-memory=5037.000000
2	avg-sec-per-sent-val=0.403712	bleu-val=0.106434	perplexity-train=40.775847	perplexity-val=203796.727398	time-elapsed=765.858212	used-gpu-memory=5037.000000
3	avg-sec-per-sent-val=0.351910	bleu-val=0.153550	perplexity-train=24.803743	perplexity-val=266022.905890	time-elapsed=1181.168308	used-gpu-memory=5037.000000
4	avg-sec-per-sent-val=0.339560	bleu-val=0.175705	perplexity-train=16.951490	perplexity-val=428458.420087	time-elapsed=1593.712781	used-gpu-memory=5037.000000
5	avg-sec-per-sent-val=0.317350	bleu-val=0.193712	perplexity-train=13.384035	perplexity-val=571508.121336	time-elapsed=1997.266849	used-gpu-memory=5037.000000
6	avg-sec-per-sent-val=0.297145	bleu-val=0.193210	perplexity-train=12.003007	perplexity-val=609632.199078	time-elapsed=2403.232710	used-gpu-memory=5037.000000
7	avg-sec-per-sent-val=0.288253	bleu-val=0.203492	perplexity-train=10.607567	perplexity-val=689141.407104	time-elapsed=2808.074823	used-gpu-memory=5037.000000
8	avg-sec-per-sent-val=0.295245	bleu-val=0.205443	perplexity-train=9.640161	perplexity-val=591580.788622	time-elapsed=3206.099968	used-gpu-memory=5037.000000
9	avg-sec-per-sent-val=0.188597	bleu-val=0.209383	perplexity-train=9.180378	perplexity-val=780487.730170	time-elapsed=3607.191386	used-gpu-memory=5037.000000
```

**Comment by fhieber** [comment]
Can you post your training log with stats about the data as well as the command line flags?
What version of MXNet are you using?

**Comment by fhieber** [comment]
#186 this should solve it, I think

**Comment by xingniu** [comment]
Yes, it now works well. Thanks!

**Comment by fhieber** [comment]
Note, that I just found another bug which is more subtle (as it still learns but does not converge to a good result): #187

**Comment by xingniu** [comment]
With fix #186 I got a trajectory like
```
1	avg-sec-per-sent-val=0.198416	bleu-val=0.076564	perplexity-train=163.603173	perplexity-val=63.072071	time-elapsed=233.372106	used-gpu-memory=4993.000000
2	avg-sec-per-sent-val=0.190757	bleu-val=0.122397	perplexity-train=40.655208	perplexity-val=33.937285	time-elapsed=493.927438	used-gpu-memory=4993.000000
3	avg-sec-per-sent-val=0.183530	bleu-val=0.155646	perplexity-train=24.919225	perplexity-val=25.796570	time-elapsed=759.624463	used-gpu-memory=4993.000000
```
But using the same setting with fix #187 (pip install 1.10.4) I got
```
1	avg-sec-per-sent-val=0.173942	bleu-val=0.093596	perplexity-train=162.532697	perplexity-val=2828.775026	time-elapsed=227.710570	used-gpu-memory=4993.000000
2	avg-sec-per-sent-val=0.178886	bleu-val=0.000141	perplexity-train=145.261739	perplexity-val=456.754609	time-elapsed=480.116986	used-gpu-memory=4993.000000
3	avg-sec-per-sent-val=0.187132	bleu-val=0.000176	perplexity-train=421.647606	perplexity-val=450.744996	time-elapsed=740.794536	used-gpu-memory=4993.000000
```
Numbers offtracked.

**Comment by fhieber** [comment]
#188 
I do hope this is the last one of these bugfixes :)
